AR: take precision into count when assigning a value to timestamp attribute

bogdan opened the pull request:
## The Problem

Consider this following on MySQL (that have 0 precision on datetime by default):

``` ruby
d1 = Developer.create!(name: 'bogdan')
d2 = Developer.find(d1.id)
d1.created_at.usec # => 837728
d2.created_at.usec # => 0
# It results in a madness:
d1.created_at == d2.created_at # => false
d1.cache_key == d2.cache_key # => false
```

Rails 4.1.10 and master are affected. Not sure about others.

Timestamp column can have less precision than ruby timestamp
In result in how big a fraction of a second can be stored in the
database.

If the precision is low enough, (mysql default is 0 so it is always low
enough by default) the value changes when model is reloaded from the
database.
## The Patch

In order to fix that we just use a precision of a column when assigning a value to the timestamp attribute as time object (not as string - that seems working).
## Fixing tests

When I've applied a patch, the new problem raised: existing tests are written incorrectly for mysql that have 0 precision by default. 
There are many tests that check that `updated_at` changes on particular model. But it is only true for ruby level but not in the database. This patch reveals the problem. 

In order to make them pass I've set the precision for tables that are used to check if `updated_at` is changing big enough and consistent for all databases.
## Further problems

`datetime` column type works differently in different databases now because if different default precision. It results in different behaviour of the same AR schema on different databases.

It should be consistent.  It can be done by setting up a `precision` option of `add_column` to some value all the time and disallowing it to be nil.

rafaelfranca commented:
Tests are broken because datetime precision is not supported at the version of mysql that we have in our servers. This make me wonder, would this patch work on old MySQL versions?

bogdan commented:
@rafaelfranca yes it will: at the rails level, unsupported precision works like `precision # =>  0`. It means that all seconds fractions will be removed. 

The fact that older mysql don't support precision raises the following problem:

``` ruby
  def test_cache_key_changes_when_child_touched
    owner = owners(:blackbeard)
    pet   = pets(:parrot)

    owner.update_column :updated_at, Time.current
    key = owner.cache_key

    assert pet.touch
    assert_not_equal key, owner.reload.cache_key
  end
```

It will only work stable if there will be at least 1 second between the creation of `owner` and `touch` of `pet`. It can be solved by putting `sleep(1)` there or by marking as pending on old versions of mysql.

I've put `sleep(1)` initially but it didn't work because too many tests would require it and they end up being very slow.

I'll think more how we can solve it. If you have an idea, please share with me.

rafaelfranca commented:
cc @sgrif

bogdan commented:
Hope I've fixed the build on mysql5.6: I made it differently with `travel` so that tests don't relay on ruby-is-slow.

Can someone advice how to go from PR to corresponding CI build? Find it here: https://travis-ci.org/rails/rails/pull_requests doesn't work.

bogdan commented:
@sgrif @rafaelfranca I understand the work load of rails core team, but this is regression problem from 4.0 to 4.1 and we kind of expect a feedback sooner than later. Can you please merge this in and backport to 4.1 or point out why it can not be done?

sgrif commented:
Forcing a Travis build.

sgrif commented:
I'm traveling and don't have my computer right now but I'll take care of this next chance I get.

bogdan commented:
@sgrif if you have back to work, it would be cool to receive another review here

sgrif commented:
Are you Github stalking me? :trollface: Taking a look now.

sgrif commented on activerecord/test/cases/timestamp_test.rb:
```diff
@@ -457,6 +464,14 @@ def test_index_is_created_for_both_timestamps
+    in_instance = d.created_at
+    from_db = Parrot.find(d.id).created_at
+    assert_equal from_db.usec, in_instance.usec
+    assert_equal from_db, in_instance
```
I find this test hard to follow, and I don't think this is relevant to the behavior of AR::Timestamp. Can we move this to a separate test in something like `test/cases/type/date_time_test.rb`, and have it look something like this?

```ruby
model = Model.new(date_time_attr_that_isnt_auto_touched: Time.now)
assert_equal model.date_time_attr, model.tap(&:save).reload.date_time_attr
```

bogdan replied:
I don't see any test like that inside `am/test/cases/type/` that uses `Model`. They all just instantiate the `AM::Type::*` object and call `#cast` method there.

Also, testing this feature in AM will not cover this change:
https://github.com/rails/rails/pull/20317/files#diff-2695559f74d901580d9c9314a9f4aff7R14
that caused postgresql specific bug.

I would stick with test we have know. I will only change it with  use of `#reload`.

sgrif replied:
If it uses a model, create a test in the active record namespace.

sgrif replied:
There are already several tests of this nature in `activerecord/test/cases/type`

sgrif commented on activerecord/lib/active_record/type/helpers/time_value.rb:
```diff
@@ -20,6 +16,13 @@ def serialize(value)
           value
         end
 
+        def apply_precision(value)
```
What do you think about naming this `apply_nanosecond_precision` instead?

sgrif replied:
Or maybe `normalize_subsecond_precision`?

sgrif commented:
A few nitpicks. This will need to be rebased. The relevant classes have been moved to Active Model.

bogdan committed:
Fixed taking precision into count when assigning a value to timestamp attribute

Timestamp column can have less precision than ruby timestamp
In result in how big a fraction of a second can be stored in the
database.

  m = Model.create!
  m.created_at.usec == m.reload.created_at.usec
    # => false
    # due to different seconds precision in Time.now and database column

If the precision is low enough, (mysql default is 0, so it is always low
enough by default) the value changes when model is reloaded from the
database. This patch fixes that issue ensuring that any timestamp
assigned as an attribute is converted to column precision under the
attribute.

bogdan commented:
You are right about that test. I didn't get what you meant from the first try.

So:

* Renamed method to `apply_seconds_precision`. Feel like it is the best.
* Reworked test to be in proper place and don't depend on autoupdated time stamps.
